dna/codon.go

```diff
@@ -1,264 +1,245 @@
 	}
-	return answer
+	for i := 0; i < len(b)-frame; i += 3 {
+		answer = append(answer, Codon{b[i], b[i+1], b[i+2]})
+	}
+	return answer, err
 }
 
-//CodonsToSeq reverts a slice of Codons into a slice of DNA bases.
-func CodonsToSeq(c []*Codon) []Base {
-	var answer []Base
-	for i := 0; i < len(c); i++ {
-		for j := 0; j < len(c[i].Seq); j++ {
-			answer = append(answer, c[i].Seq[j])
-		}
+// CodonsToSeq converts a slice of Codons into a slice of DNA bases.
+func CodonsToSeq(c []Codon) []Base {
+	answer := make([]Base, 0, len(c)*3)
+	for i := range c {
+		answer = append(answer, c[i][0], c[i][1], c[i][2])
 	}
 	return answer
 }
 
-//TranslateCodon converts an individual Codon struct into the corresponding AminoAcid type.
-func TranslateCodon(c *Codon) AminoAcid {
-	val, ok := GeneticCode[BasesToString(c.Seq)]
+//TranslateCodon converts an individual Codon into the corresponding AminoAcid type.
+func TranslateCodon(c Codon) (AminoAcid, error) {
+	val, ok := GeneticCode[c]
 	if ok {
-		return val
+		return val, nil
 	} else {
-		log.Fatalf("Codon not in map.\n")
-		return val
+		return val, ErrCodonNotInMap
 	}
 }
 
-//NonSynonymous compares two Codon structs and returns true if they encode different AminoAcids.
-func NonSynonymous(c1 *Codon, c2 *Codon) bool {
-	return TranslateCodon(c1) != TranslateCodon(c2)
+// NonSynonymous compares two Codons and returns true if they encode different AminoAcids.
+func NonSynonymous(c1 Codon, c2 Codon) bool {
+	p1, _ := TranslateCodon(c1)
+	p2, _ := TranslateCodon(c2)
+	return p1 != p2
 }
 
-//Synonymous compares two codons and returns true if the codones have non-identical sequences that code for the same amino acid.
-func Synonymous(c1 *Codon, c2 *Codon) bool {
-	return TranslateCodon(c1) == TranslateCodon(c2) && !IsEqual(c1, c2)
+// Synonymous compares two codons and returns true if the codons code for the same amino acid.
+func Synonymous(c1 Codon, c2 Codon) bool {
+	p1, _ := TranslateCodon(c1)
+	p2, _ := TranslateCodon(c2)
+	return p1 == p2
 }
 
-//IsEqual compares two Codons and returns true if tthe underlying sequences are identical.
-func IsEqual(c1 *Codon, c2 *Codon) bool {
-	return BasesToString(c1.Seq) == BasesToString(c2.Seq)
+// IsEqual compares two Codons and returns true if the underlying sequences are identical.
+func IsEqual(c1 Codon, c2 Codon) bool {
+	return c1 == c2
 }
 
-//TranslateSeq takes a sequence of DNA bases and translates it into a slice of Amino acids.
-func TranslateSeq(b []Base) []AminoAcid {
-	var answer []AminoAcid
-	codons := BasesToCodons(b)
-	for i := 0; i < len(codons); i++ {
-		answer = append(answer, TranslateCodon(codons[i]))
+// TranslateSeq takes a sequence of DNA bases and translates it into a slice of Amino acids.
+func TranslateSeq(b []Base) ([]AminoAcid, error) {
+	answer := make([]AminoAcid, len(b)/3)
+	codons, err := BasesToCodons(b)
+	for i := range codons {
+		answer[i], err = TranslateCodon(codons[i])
 	}
-	return answer
+	return answer, err
```

Comment:
I think we may have a bug in the error handling code here.  Right now it only returns the err value associated with the last call to TranslateCodon, but we ignore all previous calls to TranslateCodon as well as the call to BasesToCodons.  I see three possible ways to do it:
1) return first error encountered
2) return "worst" error encountered
3) return all errors

There would also be a few different ways to handle what gets returned for the []AminoAcid where there are errors at some locations but not others.  We could return all AminoAcids that could be translated without an error, just those up to the first error, or all that could be translated with some sort of "filler" codon in the places where there was an error.  Maybe similar to N or Gap in DNA.
I think this goes to a bigger question of what do we want the programs to do when there is an "upstream" error detected in either the input or in "upstream" code that has caused something that "does make sense" to appear at some point in the program.  I think we have a few choices:
1) Abort and make the user/programmer fix it
2) Try to process parts of the data that make sense and return the answer for those, ignoring the other parts
3) Take a best-guess as to what the parts that don't make sense might be.

There may also be other choices here that I have not thought of.  Right now we are mostly programming in the style of 1, but I am interested to hear everyone's thoughts.  2 and 3 make me nervous because I worry it can lead to either us, or people who use our code, publishing incorrect results.  I think 2 and 3 will also lead to a lot of extra thought (e.g. what should we do when we get X as input?) and extra code (implementation of handling different types of errors and doing the best we can) as well as additional testing code to make sure all the error handling is working as expected